Cleaner TileEntityItemStackRenderers

Shadows-of-Fire opened the pull request:
This PR seeks to clarify the current situation with `TileEntityItemStackRenderer`.  

As it stands, the current way to render a tile entity as an item like vanilla does, is a forge hook inside `TileEnitityItemStackRenderer` using a deprecated system within `ForgeHooksClient`.  As it is marked deprecated, and commented "Will be removed as soon as possible," this PR will allow for its complete removal as a system in 1.13.  The only reason the PR does not remove it, is to allow it to be merged within 1.12.2's lifespan.  

When the deprecated system is removed, the `Map<Item, TileEntityItemStackRenderer>` could be moved to `ForgeHooksClient` in its place.  Or it could be placed there now, but I felt it was better placed within `TileEntityItemStackRenderer` for now.  It can be moved if desired, but I would like to know where people would want it to be, if not where it is currently.

Shadows-of-Fire committed:
Cleaner TEISRs

williewillus commented:
This is just moving the map from a forge class into a patch - I don't really see the point of it?

What was meant by "remove as soon as possible" was "improve the model system to the point that TEISR's are no longer needed", which for some use cases has not happened yet.

Shadows-of-Fire commented:
I don't see that happening.  The current system that was put in place is also poorly set up, while this is a vanilla system that is aleeady in place and has far more ease of use.  

As for the map being on the patch - that can be moved, but I was looking for feedback on other possible locations for it.

williewillus commented:
The location of the map isn't the problem, it's just that I don't see how this improves on the existing system at all, other than adding another way to register the same thing (TE renderers for Items)

Shadows-of-Fire commented:
This improves over the old system mostly by directness and information.  Currently, the registration method relies on passing Item, int, and tile class, and in the end, you get to render directly with your TESR using a null tile and a lack of certain information - specifically, the ItemStack.  

That means two things.  You have to special case your TESR to render when there is a null tile, which is sub-optimal at best, and you cannot rely on information stored in the ItemStack for anything - such as vanilla does for the shield or skull.  

As for directness, this allows you to properly create a separate class to work with the ItemStack, and it has the benefit of not passing through the entire tile entity registry to find the needed TESR.

Currently, this is possible by replacing the static instance of `TileEntityItemStackRenderer` and delegating to the old one, but that is not desirable.

williewillus commented:
Makes sense

williewillus commented on patches/minecraft/net/minecraft/client/renderer/tileentity/TileEntityItemStackRenderer.java.patch:
```diff
@@ -1,6 +1,14 @@
+     private final TileEntityBed field_193843_g = new TileEntityBed();
+     private final TileEntitySkull field_179023_f = new TileEntitySkull();
+     private final ModelShield field_187318_g = new ModelShield();
++    public static final java.util.Map<Item, TileEntityItemStackRenderer> ITEM_TO_RENDERER = new java.util.HashMap<>();
```
Should use registry delegates (see patches to ItemColors/BlockColors)

....I think at least

Shadows-of-Fire committed:
Use registry delegates

mezz commented on patches/minecraft/net/minecraft/client/renderer/tileentity/TileEntityItemStackRenderer.java.patch:
```diff
@@ -1,6 +1,14 @@
+     private final TileEntityBed field_193843_g = new TileEntityBed();
+     private final TileEntitySkull field_179023_f = new TileEntitySkull();
+     private final ModelShield field_187318_g = new ModelShield();
++    public static final java.util.Map<net.minecraftforge.registries.IRegistryDelegate<Item>, TileEntityItemStackRenderer> ITEM_TO_RENDERER = new java.util.HashMap<>();
```
Since the keys are unique, I think this can be an `IdentityHashMap` to squeeze out a bit more performance.

liach replied:
You should encapsulate this field.

Shadows-of-Fire replied:
No point to make the patch bigger.  The public field is fine.

Shadows-of-Fire committed:
IdentityHashMap

Shadows-of-Fire committed:
Determined to be a bad idea

Using IdentityHashMap defeats the purpose of using IRegistryDelegate in
the first place, which is to allow overrides to retrieve the correct
render from the map.

liach commented on patches/minecraft/net/minecraft/client/renderer/RenderItem.java.patch:
```diff
@@ -9,6 +9,15 @@
+             {
+                 GlStateManager.func_179131_c(1.0F, 1.0F, 1.0F, 1.0F);
+                 GlStateManager.func_179091_B();
+-                TileEntityItemStackRenderer.field_147719_a.func_179022_a(p_180454_1_);
```
Doesn't this have one more usage in chestrenderer class?

Shadows-of-Fire committed:
Apply changes to ChestRenderer

liach commented:
can you add a test mod adding a builtin model?

Shadows-of-Fire committed:
Use a method on Item instead of a Map

Since performance was a concern with the map, an Item method appears to
be the better option.

liach commented on patches/minecraft/net/minecraft/item/Item.java.patch:
```diff
@@ -791,13 +791,21 @@
++    /**
++     * @return This Item's TEISR, this method should NOT return a new instance.  Must not be null.
++     */
++    public net.minecraft.client.renderer.tileentity.TileEntityItemStackRenderer getTEISR()
```
You should name this method `getSpecialRenderer()` to make it more clear that users ~may draw models or call tile entity renderers in this method~.

Shadows-of-Fire replied:
Users should not draw models or call tile entity renderers in this method.  They should be returning a (preferably static) TileEntityItemStackRenderer instance, which will do those things.

liach replied:
I meant that they can do that in the special renderer itself

covers1624 replied:
I agree with the name change, it should also be marked as `@SideOnly (Side.CLIENT)`

mezz replied:
`getTileEntityItemStackRenderer` would be fine by me. I don't think the abbreviation is appropriate, it's way too specific to be abbreviated casually in most contexts, and definitely not well known enough to use it formally here.

Shadows-of-Fire committed:
Change func name + sideonly

mezz commented on src/main/java/net/minecraftforge/client/ForgeHooksClient.java:
```diff
@@ -581,7 +581,7 @@ public static void renderTileItem(Item item, int metadata)
 
     /**
-     * @deprecated Will be removed as soon as possible, hopefully 1.9.
+     * @deprecated Will be removed as soon as possible, hopefully 1.9.  See {@link Item#getTEISR()}.
```
javadoc link needs to be updated

Shadows-of-Fire committed:
update jdoc

williewillus commented on patches/minecraft/net/minecraft/item/Item.java.patch:
```diff
@@ -791,13 +791,22 @@
 +    public void onHorseArmorTick(World world, net.minecraft.entity.EntityLiving horse, ItemStack armor) {}
++    
++    /**
++     * @return This Item's TEISR, this method should NOT return a new instance.  Must not be null.
```
I think this might be a bit unclear, someone could take "should not return a new instance" to mean "don't do anything custom" which would make it quite a useless method indeed.

Maybe something like "the return value of this method should be constant for the lifetime of the client" or similar

williewillus commented on src/main/java/net/minecraftforge/client/ForgeHooksClient.java:
```diff
@@ -581,7 +581,7 @@ public static void renderTileItem(Item item, int metadata)
 
     /**
-     * @deprecated Will be removed as soon as possible, hopefully 1.9.
+     * @deprecated Will be removed as soon as possible, hopefully 1.9.  See {@link Item#getTileEntityItemStackRenderer()}.
```
since we're now 3 major versions past 1.9, maybe we can remove that part of the comment lol

Shadows-of-Fire committed:
update comments

Shadows-of-Fire commented:
Just writing something here so the stale bot doesn't mark this in 3 days.

LexManos commented:
I am not a fan of this method due to how it encourages modders to dynamically make their TESR's in this function as well as teaches the Item directly about client rendering things.
Is there anything you can think of that would force modders to NOT return new instances in this method? It's very important that that does not happen.

Shadows-of-Fire commented:
I could make the method final and add a field and a setter.

Shadows-of-Fire commented:
Would it be better to have a protected or public setter for this, or would it not matter?

bs2609 commented:
Just have a protected field and no setter?

Shadows-of-Fire committed:
Add field + setter

liach commented on patches/minecraft/net/minecraft/item/Item.java.patch:
```diff
@@ -792,21 +792,30 @@
 +    public void onHorseArmorTick(World world, net.minecraft.entity.EntityLiving horse, ItemStack armor) {}
 +    
++    @SideOnly(Side.CLIENT)
++    private net.minecraft.client.renderer.tileentity.TileEntityItemStackRenderer teisr = null;
```
`@Nullable`

Shadows-of-Fire replied:
Is it necessary?  The field is initialized to be null, and it's private...  It'd be an extra line with no real purpose.

mezz replied:
IDEs will complain, might as well make it nice?

williewillus commented on patches/minecraft/net/minecraft/item/Item.java.patch:
```diff
@@ -791,13 +791,31 @@
++    private net.minecraft.client.renderer.tileentity.TileEntityItemStackRenderer teisr = null;
++    
++    /**
++     * @return This Item's renderer.  Nonnull.
```
use the annotation instead of a comment

Shadows-of-Fire replied:
The package has `@MethodsReturnNonnullByDefault`, so I don't think either are necessary, actually.

Shadows-of-Fire committed:
comment update

Shadows-of-Fire commented:
I believe these changes should be effective in forcing users to never return a new instance.  If this is not adequate, the only other solution I can see is going back to the map version in the earliest form of this PR.

mezz commented:
This looks like a good solution. I don't really see how an implementer can mess this one up.

Shadows-of-Fire committed:
Nullable

mezz commented on patches/minecraft/net/minecraft/item/Item.java.patch:
```diff
@@ -791,13 +791,32 @@
++    }
++    
++    @SideOnly(Side.CLIENT)
++    public void setTileEntityItemStackRenderer(@Nullable net.minecraft.client.renderer.tileentity.TileEntityItemStackRenderer teisr)
```
Can you make this `protected` so that you can only change your own item's renderer?

covers1624 replied:
Protected means that it would have to be called from either the package containing your item class or the class itself, i'd prefer it throw an exception if the renderer is not null.

mezz replied:
If you want different access, you can override it in your `Item` method. The idea is just to make it harder to really mess up something else's rendering.

Shadows-of-Fire committed:
Merge remote-tracking branch 'refs/remotes/MinecraftForge/1.12.x' into teisr

Conflicts to be fixed soon

Shadows-of-Fire committed:
Fix merge conflicts from last commit

Shadows-of-Fire commented:
Merge conflicts resolved.  Would love if we could get this moving forward, it's been almost three weeks since changes were approved, and this is a rather important update to this system.

LexManos commented:
Still not a fan of hard linking rendering code to the Item. And your set function doesn't really solve anything... But it works for now.

diesieben07 commented:
This breaks servers. You cannot have an initializer for a `@SideOnly` field.
See: http://www.minecraftforge.net/forum/topic/63738-server-wont-start-forge-1122-142322638/

LexManos commented:
God i knew I felt something wrong about that... @mezz should of tested dedicated server before this.

HenryLoenwind commented:
Yeah, it's annoying that the compiler actually creates code for an `= null` field initializer.

(Not getting them stripped out is also annoying, but as I understand it, there's no good way to match their code to the field.)

diesieben07 commented:
> Yeah, it's annoying that the compiler actually creates code for an = null field initializer.

Not really. It is mandated by the Java spec. If the field is final, this causes more things to happen than just "the field now has a value".

Shadows-of-Fire commented:
I didn't even realize that was a thing.  My fault.  Does simply removing the null initializer resolve the crash?

williewillus commented:
no, because class members are initialized with their default value, iirc. the `= null` will just happen implicitly

diesieben07 commented:
> no, because class members are initialized with their default value, iirc. the = null will just happen implicitly

Yes, it will. The default value is applied implicitly by the JVM, it is not generated by the Java compiler.

Shadows-of-Fire commented:
Then that would make the only viable solution to remove the safeguards by de-finalizing the getter, and delete the field + setter.

Shadows-of-Fire commented:
But I recall Lex wasn't thrilled by the idea of that

HenryLoenwind commented:
In *this* case the assignment is not needed, the field is not final. So just dropping it is fine. Methods are of no concern here.

Background:

`int a = 1;`

compiles basically to

```
int a;
static {
  a=1;
}
```

and after stripping the field, the static block still stands. For `Object x = null;` the assignment is pointless but is created anyway. For `final Object b = null;` the assignment serves a purpose, so it cannot be dropped and annotating such a field with @Optional just won't work.
